activerecord/test/cases/transaction_callbacks_test.rb

```diff
@@ -108,6 +108,19 @@ def test_call_after_commit_after_transaction_commits
     assert_equal [:after_commit], @first.history
   end
 
+  def test_do_not_call_after_commit_or_after_update_commit_after_transaction_commits_if_update_fails
+    @first.after_commit_block { |r| r.history << :after_commit }
+    @first.after_commit_block(:update) { |r| r.history << :commit_on_update }
```

Comment:
The described bug only exists for `after_commit` callbacks, but I'm testing for both `after_commit` and `after_update_commit` because I want to avoid regressions for `after_update_commit`. The reason the bug exists for one and not the other is subtle and not necessarily intentional in the current code.